sky/serve/core.py

```diff
@@ -599,6 +609,167 @@ def status(
     return serve_utils.load_service_status(serve_status_payload)
 
 
+@usage_lib.entrypoint
+def sync_down_logs(service_name: str,
+                   service_component: Optional[serve_utils.ServiceComponent],
+                   replica_id: Optional[int]) -> None:
+    """Sync down logs for a given service.
+
+    - This function synchronizes logs from remote service to local machine.
+    - It supports downloading logs for different components of the service.
+
+    Args:
+        service_name: name of service.
+        service_component: component to sync down the logs of.
+            - Could be controller, load balancer, replica, or None.
+            - None means to sync down logs for everything.
+        replica_id: The replica ID to tail the logs of.
+            - Only used when target is replica.
+
+    Raises:
+        ValueError: If the service_component is not a valid string or
+                    serve_utils.ServiceComponent, or if replica_id is not
+                    specified when service_component is REPLICA.
+        RuntimeError: If there is an error during log sync process.
+    """
+
+    if isinstance(service_component, str):
+        service_component = serve_utils.ServiceComponent(service_component)
+    if not isinstance(service_component, serve_utils.ServiceComponent):
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError(
+                f'`service_component` must be a string or '
+                f'sky.serve.ServiceComponent, got {type(service_component)}.')
+    if service_component == serve_utils.ServiceComponent.REPLICA:
+        if replica_id is None:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError('`replica_id` must be specified when '
+                                 'using service_component=REPLICA.')
+    else:
+        if replica_id is not None:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError('`replica_id` must be None when using '
+                                 'service_component=CONTROLLER/LOAD_BALANCER.')
+
+    logger.info(f'Syncing down logs for {service_name}...')
+    controller_handle = backend_utils.is_controller_accessible(
+        controller=controller_utils.Controllers.SKY_SERVE_CONTROLLER,
+        stopped_message='No service is found.')
+
+    # Prepare logs for download.
+    if service_component is None:
+        assert replica_id is None
+
+    # Get SSH credentials and initialize SSHCommandRunner.
+    ssh_credentials = backend_utils.ssh_credential_from_yaml(
+        controller_handle.cluster_yaml, controller_handle.docker_user)
+    runner = command_runner.SSHCommandRunner(
+        node=(controller_handle.head_ip, controller_handle.head_ssh_port),
+        **ssh_credentials,
+    )
```

Comment:
Got it, but I'm a little bit confused, if I move the following:

```python
# Get SSH credentials and initialize SSHCommandRunner.
ssh_credentials = backend_utils.ssh_credential_from_yaml(
    controller_handle.cluster_yaml, controller_handle.docker_user)
runner = command_runner.SSHCommandRunner(
    node=(controller_handle.head_ip, controller_handle.head_ssh_port),
    **ssh_credentials,
)
```

to the beginning of every run, this means I would have to write it three times, inside:

```python
if (sync_down_all_components or
        service_component == serve_utils.ServiceComponent.REPLICA):
```

```python
if (sync_down_all_components or
        service_component == serve_utils.ServiceComponent.CONTROLLER):
```

```python
if (sync_down_all_components or
        service_component == serve_utils.ServiceComponent.LOAD_BALANCER):
```

It may introduces some redundancy?